### PR TITLE
Add support for CPI api v2

### PR DESCRIPTION
**Pull request by andrew-su**
This change by @miguelverissimo and I affects the following calls:
- Create VM
- Attach Disk
- Info

This change also removes the legacy "registry client" as well.

Relevant story: https://www.pivotaltracker.com/story/show/163148757

### CODE REVIEW

**Comment by xtreme-andrew-su** [comment]
Currently the expectation is that both V1 and V2 is supported. When it comes time to deprecating V1 it should be updated in the release to return an appropriate error message for deprecated versions.

As per maintainability, I agree that's not a very clean way of doing that. We can revisit how we choose the right api.

I've created a story for the update to the PR: https://www.pivotaltracker.com/story/show/163483524

**Comment by evandbrown** [comment]
Thanks, Andrew. What's the relative priority here? I'm on vacation for the
next 10 days but can test and merge this when I'm back. Sound OK?

**Comment by xtreme-andrew-su** [comment]
It should be fine to wait until you're back from your vacation. Enjoy!

**Comment by xtreme-andrew-su** [comment]
Hello @evandbrown, have you had a chance to take a look at the changes?

**Comment by h4xnoodle** [comment]
@evandbrown following from Andrew's comment, do you think we could review this soon? Thank you!

File: src/bosh-google-cpi/action/factory_interface.go [hunk 1]
```diff
@@ -1,5 +1,5 @@
 package action
 
 type Factory interface {
-	Create(method string, ctx map[string]interface{}) (Action, error)
+	Create(method string, ctx map[string]interface{}, apiVersion int) (Action, error)
```
Comment (evandbrown) [inline src/bosh-google-cpi/action/factory_interface.go]: Are API versions guaranteed to be ints?

File: src/bosh-google-cpi/action/concrete_factory.go [hunk 2]
```diff
@@ -127,22 +127,12 @@ func (f ConcreteFactory) Create(method string, ctx map[string]interface{}) (Acti
...
-	// value of ClientOptions.UseGCEMetadata
-	var registryClient registry.Client
-	switch f.cfg.Cloud.Properties.Registry.UseGCEMetadata {
-	case true:
-		registryClient = registry.NewMetadataClient(
-			googleClient,
-			f.cfg.Cloud.Properties.Registry,
-			f.logger,
-		)
-	default:
```
Comment (evandbrown) [inline src/bosh-google-cpi/action/concrete_factory.go]: Is there a reason to remove support for the HTTP client?

File: src/bosh-google-cpi/action/create_vm_test.go [hunk 3]
```diff
@@ -205,23 +214,50 @@ var _ = Describe("CreateVM", func() {
...
-			Expect(diskTypeService.FindCalled).To(BeFalse())
-			Expect(vmService.CreateCalled).To(BeTrue())
-			Expect(vmService.CleanUpCalled).To(BeFalse())
-			Expect(registryClient.UpdateCalled).To(BeTrue())
-			Expect(registryClient.UpdateSettings).To(Equal(expectedAgentSettings))
-			Expect(vmCID).To(Equal(VMCID("fake-vm-id")))
-			Expect(vmService.CreateVMProps).To(Equal(expectedVMProps))
-			Expect(vmService.CreateNetworks).To(Equal(expectedInstanceNetworks))
-			Expect(vmService.CreateRegistryEndpoint).To(Equal("http://fake-registry-username:fake-registry-password@fake-registry-host:25777"))
+		Context("with aoi v1", func() {
```
[the remark this answers is not in the document]
Comment (andrew-su) [reply src/bosh-google-cpi/action/create_vm_test.go]: I couldn't find this typo.

File: src/bosh-google-cpi/action/factory_interface.go (hunk 1 above)
Comment (andrew-su) [reply src/bosh-google-cpi/action/factory_interface.go]: AFAIK, there's no plans to change it from ints.

File: src/bosh-google-cpi/action/concrete_factory.go (hunk 2 above)
Comment (andrew-su) [reply src/bosh-google-cpi/action/concrete_factory.go]: Although this isn't "deadcode" per se, but the release itself offers no way for an operator to use it.

The configurable option `registry.use_gce_metadata` in `jobs/google_cpi/spec` doesn't get used in the `cpi.json.erb`; it gets defaulted to `true`.

I don't recall tests using this either, please correct me if I'm mistaken.

File: src/bosh-google-cpi/action/create_vm_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrew-su) [reply src/bosh-google-cpi/action/create_vm_test.go]: Found it, but it was fixed in the latest commit 39f5c62.

**Pull request merged**